pkg/common/git.go

```diff
@@ -0,0 +1,192 @@
+package common
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/nuclio/nuclio/pkg/cmdrunner"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	githttp "gopkg.in/src-d/go-git.v4/plumbing/transport/http"
+)
+
+type GitAttributes struct {
+	Branch    string `json:"branch,omitempty"`
+	Tag       string `json:"tag,omitempty"`
+	Reference string `json:"reference,omitempty"`
+	Username  string `json:"username,omitempty"`
+	Password  string `json:"password,omitempty"`
+}
+
+type GitClient interface {
+	Clone(outputDir, repositoryURL string, gitAttributes *GitAttributes) error
+}
+
+type AbstractGitClient struct {
+	GitClient
+
+	logger    logger.Logger
+	cmdRunner cmdrunner.CmdRunner
+}
+
+func NewGitClient(parentLogger logger.Logger) (GitClient, error) {
+	var err error
+
+	abstractGitClient := AbstractGitClient{logger: parentLogger.GetChild("git-client")}
+
+	// create cmd runner
+	abstractGitClient.cmdRunner, err = cmdrunner.NewShellRunner(parentLogger)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create cmd runner")
+	}
+
+	return abstractGitClient, nil
+}
+
+func (agc AbstractGitClient) Clone(outputDir, repositoryURL string, gitAttributes *GitAttributes) error {
+	var referenceName string
+	var gitAuth *githttp.BasicAuth
+	var err error
+
+	// resolve full git reference name
+	referenceName, err = ResolveGitReference(repositoryURL, gitAttributes)
+	if err != nil {
+		return errors.Wrap(err, "Failed to resolve git reference")
+	}
+
+	// resolve git credentials when given
+	gitAuth = agc.parseFunctionGitCredentials(gitAttributes)
+
+	// if it's Azure Devops repo - clone differently (the normal go-git client doesn't support it yet)
```

Comment:
afaik they have an github issue ticket for that, add that to comment
+ add `// TODO:` for easier comments lookup